KEP-5142: Pop pod from backoffQ when activeQ is empty (#5144)
kubernetes/enhancements: keps/sig-scheduling/5142-pop-backoffq-when-activeq-empty/README.md

- [ ] (R) Production readiness review completed
- [ ] (R) Production readiness review approved
- [ ] "Implementation History" section is up-to-date for milestone
- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

This KEP proposes improving scheduling queue behavior by popping pods from backoffQ when activeQ is empty. 
This would allow to increase utilization of kube-scheduler cycles as well as reduce waiting time for pending pods 
that were previously unschedulable.

## Motivation

> 

Reviewer: Can you add a bit more about what is backoffQ (= they put penalty on pods that wasted the scheduling cycles) and why we can pop from there (= because they can be retried but they're just waiting for a penalty).

Author: I added more context about the queues

- `scheduler_pod_scheduling_sli_duration_seconds` metric is visibly higher for schedulable pods.

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

No. This feature is a in-memory feature of the scheduler 
and thus calculations start from the beginning every time the scheduler is restarted.
So, just upgrading it and upgrade->downgrade->upgrade are both the same.

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

No

### Monitoring Requirements

###### How can an operator determine if the feature is in use by workloads?

This feature is used during scheduling when activeQ is empty and if the feature gate is enabled.

###### How can someone using this feature know that it is working for their instance?

N/A

###### What are the reasonable SLOs (Service Level Objectives) for the enhancement?

In the default scheduler, we should see the throughput around 100-150 pods/s ([ref](https://perf-dash.k8s.io/#/?jobname=gce-5000Nodes&metriccategoryname=Scheduler&metricname=LoadSchedulingThroughput&TestName=load)), and this feature shouldn't bring any regression there.

> Based on that `schedule_attempts_total` shouldn't be less than 100 in a second.

Reviewer: nit:

```suggestion
Based on that `schedule_attempts_total` shouldn't be less than 100 in a second
if there are enough pods that are constantly created within the cluster.
```

However, when activeQ is empty, kube-scheduler idles, waiting for any pod being present in activeQ, 
even if pods are in the backoffQ but their backoff period hasn't expired. 
In scenarios when pods are waiting, but in backoffQ, 
kube-scheduler should be able to consider those pods for scheduling, even if the backoff is not completed, to avoid the idle time.

### Goals

- Improve scheduling throughput and kube-scheduler utilization when activeQ is empty, but pods are waiting in backoffQ.
- Run `PreEnqueue` plugins when putting pod into backoffQ.

### Non-Goals

- Refactor scheduling queue by changing backoff logic or merging activeQ with backoffQ.

## Proposal

At the beginning of scheduling cycle, pod is popped from activeQ. 
If activeQ is empty, it waits until a pod is placed into the queue.
This KEP proposes to pop the pod from backoffQ when activeQ is empty.

To ensure the `PreEnqueue` is called for each pod taken into scheduling cycle,
`PreEnqueue` plugins would be called before putting pods into backoffQ.
It won't be done again when moving pods from backoffQ to activeQ.

### Risks and Mitigations

> #### Scheduling throughput might be affected
> 
> While popping from backoffQ, another pod might appear in activeQ ready to be scheduled.
> If the pop operation is short enough, there won't be a visible downgrade in throughput.
> The only concern might be that less pods from activeQ might be taken in some period of time in favor of backoffQ, 
> but that's a user responsibility to create enough amount of pods to be scheduled from activeQ, not to cause this KEP behavior to happen.

Reviewer: ```suggestion
#### A tiny delay on the first scheduling attempts for newly created pods

While the scheduler handles a pod directly popping from backoffQ, another pod might appear in activeQ, which should have been scheduled before a pod being scheduled now.
However, in the real world, if the scheduling latency is short enough, there won't be a visible downgrade in throughput.
```

I don't get what you're trying to mean at the last parts L145-L146. Feel free to bring those back with some elaboration.

Author: Applied your suggestion. I wanted to mention there that delay won't happen if someone just create enough pods that will be in activeQ 😅

#### Backoff won't be working as natural rate limiter in case of errors

In case of API calls errors (e.g. network issues), backoffQ allows to limit number of retries in a short term.
This proposal will take those pods earlier, leading to losing this mechanism.

> After merging [kubernetes#128748](github.com/kubernetes/kubernetes/pull/128748), 

Reviewer: We should clarify when to do this mitigation, like by alpha or by beta.

Author: Let's mitigate it until beta. Added

it will be possible to distinguish pods backing off because of errors from those backing off because of unschedulable attempt. 
This information could be used when popping, by filtering only the pods that are from unschedulable attempt or even splitting backoffQ. 

#### One pod in backoffQ could starve the others

If a pod popped from the backoffQ fails its scheduling attempt and come back to the queue, it might be selected again, ahead of other pods.

To prevent this, while popping pod from backoffQ, its attempt counter will be incremented as if it had been taken from the activeQ.

> This will give other pods a chance to be scheduled.

Reviewer: I don't get what you try to regard as an issue here. Even if we increment an attempt counter, as long as activeQ is empty, the attempt counter is ignored and the pod could go to activeQ without backoff, right? The pod doesn't block others because "others" would also be poped directly from backoffQ next. 
Can you possibly elaborate the explanation here?

Reviewer: Hmm, maybe we can just remove this section.. It sounds kinda obvious to increment the scheduling attempt count for those pods directly from backoffQ as well.

At least, I'd rephrase it like:

```suggestion
#### One pod in backoffQ could starve the others

The head of BackoffQ is the pod with the closest backoff expiration,
and the backoff time is calculated based on the number of scheduling failures that the pod has experienced.
If one pod has a smaller attempt counter than others,
might the scheduler keep popping this pod ahead of other pods because the pod's backoff expires faster than others?

Actually, that wouldn't happen because the scheduler would increment the attempt counter of pods from backoffQ as well,
which would make the backoff time of pods bigger every after the scheduling attempt,
and one pod that had a smaller attempt number eventually won't be popped out.
```

Author: I wanted to mention it because @dom4ha brought it in the issue [comment](https://github.com/kubernetes/kubernetes/issues/129806#issuecomment-2637244490). 

I applied your suggestion

## Design Details

### Popping from backoffQ in activeQ's pop()

To achieve the goal, activeQ's `pop()` method needs to be changed:

> 1. If activeQ is empty, then instead of waiting on condition, popping from backoffQ is tried.

Reviewer: ```suggestion
1. If activeQ is empty, then instead of waiting for a pod to arrive at activeQ, popping from backoffQ is tried.
```

> 2. If backoffQ is empty, then `pop()` is waiting on condition as previously.

Reviewer: ```suggestion
2. If backoffQ is empty, then `pop()` is waiting for activeQ pods as previously.
```

##### e2e tests

Feature is scoped within kube-scheduler internally, so there is no interaction between other components.
Whole feature should be already covered by integration tests.

### Graduation Criteria

#### Alpha

- Feature implemented behind a feature flag.
- All tests from [Test Plan](#test-plan) implemented.

#### Beta

- Gather feedback from users and fix reported bugs.
- Change the feature flag to be enabled by default.

#### GA

- Gather feedback from users and fix reported bugs.

### Upgrade / Downgrade Strategy

**Upgrade**

During the alpha period, users have to enable the feature gate `PopBackoffQWhenEmptyActiveQ` to opt in this feature.
This is purely in-memory feature for kube-scheduler, so no special actions are required outside the scheduler.

**Downgrade**

Users need to disable the feature gate.

### Version Skew Strategy

This is purely in-memory feature for kube-scheduler, and hence no version skew strategy.

## Production Readiness Review Questionnaire

### Feature Enablement and Rollback

###### How can this feature be enabled / disabled in a live cluster?

- [x] Feature gate (also fill in values in `kep.yaml`)

>   - Feature gate name: `PopBackoffQWhenEmptyActiveQ`

Author: Good idea, I went with the second proposal

###### What specific metrics should inform a rollback?

Abnormal values of metrics related to scheduling queue, meaning pods are stuck in activeQ:
- `scheduler_schedule_attempts_total` metric with `scheduled` label is almost constant, while there are pending pods that should be schedulable. 
  This could mean that pods from backoffQ are taken instead of those from activeQ.
- `scheduler_pending_pods` metric with `active` label is not decreasing, while with `backoff` is almost constant.
- `scheduler_pod_scheduling_sli_duration_seconds` metric is visibly higher for schedulable pods.

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

No. This feature is a in-memory feature of the scheduler 
and thus calculations start from the beginning every time the scheduler is restarted.
So, just upgrading it and upgrade->downgrade->upgrade are both the same.

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

No

### Monitoring Requirements

###### How can an operator determine if the feature is in use by workloads?

> This feature is used during scheduling when activeQ is empty and if the feature gate is enabled.

Author: That's right. I think we can use this metric. Mentioned it in KEP

- Run `PreEnqueue` plugins when putting pod into backoffQ.

### Non-Goals

- Refactor scheduling queue by changing backoff logic or merging activeQ with backoffQ.

## Proposal

At the beginning of scheduling cycle, pod is popped from activeQ. 
If activeQ is empty, it waits until a pod is placed into the queue.
This KEP proposes to pop the pod from backoffQ when activeQ is empty.

To ensure the `PreEnqueue` is called for each pod taken into scheduling cycle,
`PreEnqueue` plugins would be called before putting pods into backoffQ.
It won't be done again when moving pods from backoffQ to activeQ.

### Risks and Mitigations

#### A tiny delay on the first scheduling attempts for newly created pods

While the scheduler handles a pod directly popping from backoffQ, another pod that should be scheduled before the pod being scheduled now, may appear in activeQ.
However, in the real world, if the scheduling latency is short enough, there won't be a visible downgrade in throughput.
This will only happen if there are no pods in activeQ, so this can be mitigated by an appropriate rate of pod creation.

> #### Backoff won't be working as natural rate limiter in case of errors

Author: Right, especially if we decide to start from beta directly, it will be implemented in the first version.

This information could be used when popping, by filtering only the pods that are from unschedulable attempt or even splitting backoffQ.

This has to be resolved before the beta is released.

#### One pod in backoffQ could starve the others

The head of BackoffQ is the pod with the closest backoff expiration,
and the backoff time is calculated based on the number of scheduling failures that the pod has experienced.
If one pod has a smaller attempt counter than others,
could the scheduler keep popping this pod ahead of other pods because the pod's backoff expires faster than others?
Actually, that wouldn't happen because the scheduler would increment the attempt counter of pods from backoffQ as well,
which would make the backoff time of pods bigger every after the scheduling attempt,
and the pod that had a smaller attempt number eventually won't be popped out.

## Design Details

### Popping from backoffQ in activeQ's pop()

To achieve the goal, activeQ's `pop()` method needs to be changed:
1. If activeQ is empty, then instead of waiting for a pod to arrive at activeQ, popping from backoffQ is tried.

> 2. If backoffQ is empty, then `pop()` is waiting for pod as previously.

Author: It will be waiting for pods both from activeQ and backoffQ, whatever will be first.

> 3. If backoffQ is not empty, then the pod is processed like the pod would be taken from activeQ, including increasing attempts number.

Reviewer: >  It can cause rapid increase in the backoff time for some pods (when many event trigger retry)

I believe we should increment the attempt number. 
It is exactly an intentional behavior to move the pods to backoffQ immediately and make a lot of scheduling retries (regardless of whether with or without this feature), if many events are coming and each triggers the pod's requeueing.
As long as QHints determine the event might make the pod schedulable and then move this pod to backoffQ, this Pod should be able to get retried at any timing, even with or without backoff time. 
In case this tons of retries don't make the pod schedulable after all, and make the pod's backoff time much longer, that's the fault of QHint accuracy, not this feature.

> 3. If backoffQ is not empty, then the pod is processed like the pod would be taken from activeQ, including increasing attempts number.

Reviewer 2: We had a followup discussion with @macsko, so this is what we came up with:

We can count errors and unschedulable attempts separately, but we need two queues - for errors and for unschedulable pods, since the feature applies only to unschedulable pods (scheduling errors need the original backoff logic for rate limiting).

We agreed that it may be fine to increment attempts on every pop (we need to be prepared to see much higher values for attempts counter than before), although there are a few important things:

1. the interesting is the worst case (when active queue is permanently non-empty), so we think the existing mechanism should still apply and move expired items to the active queue - to sort them by priority.

2. since the original mechanism was triggered every 1s, the pods within this second got sorted in the active queue by priority. If we pop just based on expiration time, we would loose this, so we could have sorting by priority within every full second to reflect the current behavior.

Unfortunately I saw a case where max backoff of 10s was constantly moving big number of high priority pods to active, which were processed longer than these 10s, causing starvation of lower priority pods. So this feature does not help in this scenario directly, but it does indirectly, because we should be able to safely increase the default value to something like 120s without risking any pod would need to wait that long in practice.

> 3. If backoffQ is not empty, then the pod is processed like the pod would be taken from activeQ, including increasing attempts number.

Reviewer: > since the original mechanism was triggered every 1s, the pods within this second got sorted in the active queue by priority. 

Yeah, this is the point I missed. But,

> so we think the existing mechanism should still apply and move expired items to the active queue - to sort them by priority.

do you mean, instead of just directly popping pods out of backoffQ when activeQ is empty, we'd once move pods from backoffQ to activeQ, let activeQ sort them, and restart popping from activeQ? I.e., when activeQ is empty, trigger `flushBackoffQCompleted` forcibly?

If Yes, that means when activeQ is empty, every pop would have to iterate all pods in backoffQ (even if there're no expired items there), which looks like heavy cost.

> 3. If backoffQ is not empty, then the pod is processed like the pod would be taken from activeQ, including increasing attempts number.

Reviewer: > since the original mechanism was triggered every 1s, the pods within this second got sorted in the active queue by priority.

I'm actually not sure whether we should bring this into this feature if that would require a huge cost or further complicated implementation. 

So, the backoff flush happens every 1 sec, and there could be pods that completed backoff during this interval 1 sec. That means that pods need to wait for `{their backoff time} + ~1 sec`. This is the current scheduler's behavior.
So, let's say there are pods that completed backoff already, and among them, there are some lower priority pods and some higher priority pods. Then, even if the top pod at backoffQ among them is the lower priority pod and this feature pops the pod, either way, all pods completed backoff will be moved to activeQ at the next flushing at least, which will happen in 1 sec.
So, all pods need to wait for `{their backoff time} + ~1 sec` at backoffQ, which is the same as the current scheduler, and this feature makes some _lucky_ pods that can shortcut `+ ~1 sec` part if they're luckily at around the top of backoffQ when activeQ is empty.

> 3. If backoffQ is not empty, then the pod is processed like the pod would be taken from activeQ, including increasing attempts number.

Author: > do you mean, instead of just directly popping pods out of backoffQ when activeQ is empty, we'd once move pods from backoffQ to activeQ, let activeQ sort them, and restart popping from activeQ? I.e., when activeQ is empty, trigger `flushBackoffQCompleted` forcibly?

He just emphasized that we'll still keep the old flushing from backoffQ to activeQ apart from new popping. 

Sorting by priority for popping when activeQ is empty can be done by changing sorting key of backoffQ's heap - we could ignore the milliseconds in backoff time expiration and add priority to comparison. It would probably also require to align those windows with flushing windows, so to trigger the flush every full second.

> 3. If backoffQ is not empty, then the pod is processed like the pod would be taken from activeQ, including increasing attempts number.

Author: > So, like sorting by the expiration time first (**but ignore miliseconds**), and sorting by the priority next, right?

Exactly, I added it to the kep, so please take a look.

>    It is poping from a heap data structure, so it should be fast enough not to cause any performance troubles.

Reviewer 3: nit
```suggestion
   It is popping from a heap data structure, so it should be fast enough not to cause any performance troubles.
```

To support monitoring, when popping from backoffQ, 

> the `scheduler_queue_incoming_pods_total` metric with an `activeQ` queue and a new `PopFromBackoffQ` event label will be incremented.

Author: This metric shows the pod movement between queues, where the queue means the destination. As pod is already in backoffQ, its destination have to be activeQ (we just skip putting the pod there, but treat it as it would be in activeQ).

### Notifying activeQ condition when new pod appears in backoffQ

Pods might appear in backoffQ while `pop()` is hanging on point 2. 
That's why it will be required to call `broadcast()` on condition after adding a pod to backoffQ.
It shouldn't cause any performance issues.

We could eventually want to move backoffQ under activeQ's lock, but it's out of scope of this KEP.

### Calling PreEnqueue for backoffQ

`PreEnqueue` plugins have to be called for every pod before they are taken to scheduling cycle.
Initially, those plugins were called before moving pod to activeQ.

> With this proposal, `PreEnqueue` will need to be called before moving pod to backoffQ 
> and those calls need to be skipped for the pods that are moved later from backoffQ to activeQ.
> At moveToActiveQ level, these two paths could be distinguished by checking if event is equal to `BackoffComplete`.

Reviewer: let's mention the motivation of this change

```suggestion
Currently, we call `PreEnqueue` at every time pods are being moved to activeQ. 
But, with this proposal, `PreEnqueue` will be called before moving pod to backoffQ, not when popping pods directly from backoffQ.
Otherwise, a direct popping would be inefficient; it has to take the top backoffQ pod, check if it goes through PreEnqueue plugins, if not check the next backoffQ pod... until it finds the pod that goes through all PreEnqueue plugins.

It means we'd have two paths that PreEnqueue plugins are invoked; when new pods are created and entering to the scheduling queue and when pods are pushed in backoffQ.
At moveToActiveQ level, these two paths could be distinguished by checking if event is equal to `BackoffComplete`.
```

### Rollout, Upgrade and Rollback Planning

<!--
This section must be completed when targeting beta to a release.
-->

###### How can a rollout or rollback fail? Can it impact already running workloads?

The partly failure in the rollout isn't there because the scheduler is the only component to rollout this feature. 
But, if upgrading the scheduler itself fails somehow, new Pods won't be scheduled anymore,
while Pods, which are already scheduled, won't be affected in any case.

###### What specific metrics should inform a rollback?

Abnormal values of metrics related to scheduling queue, meaning pods are stuck in activeQ:
- `scheduler_schedule_attempts_total` metric with `scheduled` label is almost constant, while there are pending pods that should be schedulable. 
  This could mean that pods from backoffQ are taken instead of those from activeQ.
- `scheduler_pending_pods` metric with `active` label is not decreasing, while with `backoff` is almost constant.
- `scheduler_queue_incoming_pods_total` metric with `PopFromBackoffQ` label is increasing when there are pods in activeQ.
- `scheduler_pod_scheduling_sli_duration_seconds` metric is visibly higher for schedulable pods.

> 

Reviewer: Can we detect bugs potentially by checking if `scheduler_queue_incoming_pods_total` with `PopFromBackoffQ` is _always_ higher than other `event` labels? (i.e., apparently too many pods are moved by this feature always

which was the original reason of introducing the backoffQ.

To ensure the `PreEnqueue` is called for each pod taken into the scheduling cycle,
`PreEnqueue` plugins would be called before putting pods into the backoffQ.
It won't be done again when moving pods from the backoffQ to the activeQ.

### Risks and Mitigations

#### A tiny delay on the first scheduling attempts for newly created pods

While the scheduler handles a pod directly popping from the backoffQ, another pod that should be scheduled before the pod being scheduled now, may appear in the activeQ.
However, in the real world, if the scheduling latency is short enough, there won't be a visible downgrade in throughput.
This will only happen if there are no pods in the activeQ, so this can be mitigated by an appropriate rate of pod creation.

#### Backoff won't be working as a natural rate limiter in case of errors

In case of API calls errors (e.g., network issues), the backoffQ allows to limit the number of retries in a short term.
This proposal will take those pods earlier, leading to losing this mechanism.

After merging [kubernetes#128748](github.com/kubernetes/kubernetes/pull/128748),

> it will be possible to distinguish pods backing off because of errors from those backing off because of an unschedulable attempt.

Reviewer 4: Forgive me my ignorance, but if pods were unschedulable, wouldn't it be put into the `unschedulableQ` instead of `backoffQ`?
Also - if it was unschedulable, the we need some cluster state change to even wish it could become schedulable.

Reviewer: You're right for pods that were rejected by Unschedulable status. 
But, for example, when a network error happened and the pod is rejected, in this case, the pod would get Error status, and directly put in backoffQ, skipping unschedQ. This is because we cannot see when to requeue this pod (i.e., when the network failure is solved) by the cluster event. So, we just put such pod to backoffQ directly to keep it in the scheduling queue for a while, and retry it afterwards with the hope that the network failure is gone now.

What here is saying is that it's better not to skip backoff of such pods that got Error status. Because the sooner we retry those pods, the more likely the network instability is not yet solved. We want to skip backoff time of pods that were in the queue due to Unschedulable status.

Reviewer 4: > What here is saying is that it's better not to skip backoff of such pods that got Error status. Because the sooner we retry those pods, the more likely the network instability is not yet solved.

Agree with that.

> We want to skip backoff time of pods that were in the queue due to Unschedulable status.

This is what I don't understand. Aren't those pods supposed to live in `unschedulableQ` instead of `backoffQ`? If they are unschedulable, I thought they are added to `unschedulableQ` instead.

Reviewer: Those pods are put in unschedQ first, and then events trigger requeueing to backoffQ.
e.g., If a pod is rejected by the resource fit plugin, it is, first, put in unschedQ. Then, let's say new node is created, and the scheduler notices it. The queue would move this pod from unschedQ to backoffQ.
So, backoffQ could have those two types of pods; pods that got rejected by Error status, and pods that got rejected by Unschedulable status an moved from unschedQ by some event.

Reviewer 4: > Then, let's say new node is created, and the scheduler notices it. The queue would move this pod from unschedQ to backoffQ.

Ok - that was the bit I was missing. This makes sense now - thanks!

Reviewer 2: The reason why there is transition from the unschedulableQ -> backoffQ (instead of going to activeQ directly) is that the mechanism of triggering this transition not super precise for performance reasons. So it triggers whenever there is a chance the received cluster event may make an unschedulable pod schedulable, but scheduler may very often assess the pod unschedulable again, so we have so called wasted cycle.

So the backoffQ serve the purpose of a buffer so that scheduling reattempts won't consume all scheduler cycles (and starve out schedulable pods), however, at the same time it introduces scheduling delay if an event indeed made a pod schedulable. This is why we propose to pop from the backoff queue whenever activeQ is empty (the scheduler is idle anyway)

Reviewer 4: That makes sense - thanks!

This information could be used when popping, by filtering only the pods that are from an unschedulable attempt or even splitting the backoffQ.

This has to be resolved before the beta is released, which means before the release of the feature.

#### One pod in the backoffQ could starve the others

The head of the BackoffQ is the pod with the closest backoff expiration,
and the backoff time is calculated based on the number of scheduling failures that the pod has experienced.
If one pod has a smaller attempt counter than others,
could the scheduler keep popping this pod ahead of other pods because the pod's backoff expires faster than others?
Actually, that wouldn't happen because the scheduler would increment the attempt counter of pods from the backoffQ as well,

> which would make the backoff time of pods bigger every after the scheduling attempt,

Reviewer 3: nit
```suggestion
which would make the backoff time larger after each scheduling attempt,
```

When the activeQ is not empty, the scheduler pops the highest priority pod from the activeQ.
However, when the activeQ is empty, the kube-scheduler idles,
even if pods are in the backoffQ waiting for their backoff period to expire.
To avoid delaying assessment of potentially schedulable pods,
kube-scheduler should consider those pods for scheduling, even if the backoff time hasn't expired yet.
However, pods that are in the backoffQ due to errors, should not bypass the backoff time,
since it plays also rate limiting role, avoiding system overload due to too frequent retries.

### Goals

- Improve scheduling throughput and kube-scheduler utilization when the activeQ is empty, but pods are waiting in the backoffQ.
- Run `PreEnqueue` plugins when putting a pod into the backoffQ.

### Non-Goals

- Refactor the scheduling queue by changing backoff logic or merging the activeQ with the backoffQ.

## Proposal

At the beginning of the scheduling cycle, a pod is popped from the activeQ.
Currently, when activeQ is empty, it waits until some pod is placed into the queue.
This KEP proposes to pop the pod from the backoffQ when the activeQ is empty,

> however moving pods from the backoffQ to activeQ will still work as before to avoid the problem of pods starvation, 

Reviewer 2: ```suggestion
however the current mechanism of moving pods from the backoffQ to activeQ (aka flushing) will still work as before to avoid the problem of pods starvation, 
```

To preserve the efficiency of the pop() function, it will be necessary to divide the backoffQ into two queues: 
one for pods that were unschedulable, and another for those rejected due to an error.
Then popping will be performed only from the former, keeping the error backoff intact.

This has to be resolved before the beta is released, which means before the release of the feature.

#### One pod in the backoffQ could starve the others

The head of the BackoffQ is the pod with the closest backoff expiration,
and the backoff time is calculated based on the number of scheduling failures that the pod has experienced.
If one pod has a smaller attempt counter than others,
could the scheduler keep popping this pod ahead of other pods because the pod's backoff expires faster than others?
Actually, that wouldn't happen because the scheduler would increment the attempt counter of pods from the backoffQ as well,
which would make the backoff time larger after each scheduling attempt,
and the pod that had a smaller attempt number eventually won't be popped out.

### Low priority pod could be chosen to pop, even if high priority pod has a slightly later backoff expiration

> Flushing from backoffQ to activeQ is done each second, taking all pods with backoff expired.

Reviewer 2: ```suggestion
The current mechanism of flushing from backoffQ to activeQ is done each second, taking all pods with backoff expired.
```
